### PR TITLE
ensure all browsers are closed on testcafe stop

### PR DESCRIPTION
**Pull request by link89**

### CODE REVIEW

**Comment by kyle-johnson** [comment]
Thanks, the browser closing is super helpful!

I think the heartbeat has issues from time-to-time, as I get this error which I wasn't getting prior:

```
2) Unhandled promise rejection:

      UnsupportedOperationError: Error 404: Not Found
      Not Found
      at parseHttpResponse (/Users/kylejohnson/selenium-testcafe-aws/node_modules/selenium-webdriver/lib/http.js:557:11)
      at Executor.execute (/Users/kylejohnson/selenium-testcafe-aws/node_modules/selenium-webdriver/lib/http.js:468:26)
      at process._tickCallback (internal/process/next_tick.js:68:7)

**Comment by link89** [comment]
The reason I add heartbeat to keep session is we found that if testcafe exit abnormally, the browser in selenium-grid will hang for a long time until timeout is triggered. We set browser timeout to 30s to work around this problem. Since testcafe won't send any selenium command during its execution, so we implement this heartbeat login to send getTitle command every 10s to keep session. It works well in our pipeline. I think the error message you post seems like a problem of selenium: https://github.com/SeleniumHQ/selenium/issues/5784

**Comment by alexschwantes** [comment]
@link89 Looks good! +1 for fixing closing browsers.
my only thought about the heartbeat would be to potentially allow it to be configurable on/off in the event that it did cause an issue.

@kyle-johnson what browser do you experience that issue with, is it IE like in the linked selenium issue?

**Comment by link89** [comment]
@alexschwantes Your suggestion sounds reasonable. I push another commit to disable sending heartbeat by setting SELENIUM_HEARTBEAT negative.

**Comment by link89** [comment]
@kyle-johnson We also running on IE11 and Edge but didn't found the error you have occurred (driver version: 3.14.0, win10). It seems like the webdriver didn't support getTitle command. Currently I choose getTitle command as heartbeat signal, maybe you can try getCurrentUrl command to seen if the error still exists.

**Comment by kyle-johnson** [comment]
IE 11, Windows 10 (Server 2016). Selenium / IE driver version 3.14.0 (selenium run as standalone).

I think the issue is line 30 of src/indjex.js is missing an await:
`browser.get(pageUrl);`

As soon as I add an `await` in front of that, I never see the 404 error. If I leave off the await, I see the 404 sporadically, but it get it nearly every single run if I lower the heartbeat interval to 3s. I guess depending on latency, the first `getTitle` call can happen before the browser is even opened.

With the `await`, the heartbeat works like a charm (sure enough, our longer tests fail without it as selenium shuts down the browser)! 👍

**Comment by link89** [comment]
@kyle-johnson I push a commit according to your fix.

**Comment by link89** [comment]
@alexschwantes maybe we should make a new release based on recently fix, so that we don't have to maintain packages by ourselves.

File: src/index.js [hunk 1]
```diff
@@ -25,12 +27,41 @@ export default {
...
+        }
+    },
+
+    stopHeartbeat (id) {
+        this.heartbeatHandler[id] = false;
     },
 
     async closeBrowser (id) {
-        this.openedBrowsers[id].quit();
+        this.stopHeartbeat(id);
```
Comment (alexschwantes) [inline src/index.js]: add `if (this.heartbeatInterval > 0)`

File: src/index.js [hunk 2]
```diff
@@ -40,7 +71,16 @@ export default {
     },
 
     async dispose () {
-        return;
+        // ensure every session is closed on process exit
+        for (const id in this.openedBrowsers) {
+            this.stopHeartbeat(id);
```
Comment (alexschwantes) [inline src/index.js]: add `if (this.heartbeatInterval > 0)`

**Comment by alexschwantes** [comment]
after this I'll do a release ASAP. Sorry for the delay. I wasn't getting email updates of comments, but have fixed that now :)

**Comment by link89** [comment]
@alexschwantes Thanks to point this out. I have made a commit according to your suggestion.

**Pull request merged**